### ISSUE TITLE
[rllib] How does the use_lstm option work?

### ISSUE THREAD

**Issue by pschafhalter** [ray-project/ray#2536]
I'm curious about the `use_lstm` option and the associated `max_seq_len` and `lstm_cell_size` options when configuring a model in RLlib. Specifically, how does the LSTM work and what values does it remember?

The [documentation](https://ray.readthedocs.io/en/latest/rllib-models.html?highlight=lstm) says that an LSTM cell further processes the original model's output which adds a [linear layer](https://ray.readthedocs.io/en/latest/rllib-package-ref.html?highlight=lstm) at the end for final output. The diagram below illustrates my understanding:
```
state -> model -> LSTM -> action
```

Is my understanding of the LSTM correct? How does an LSTM at the output of the original model improve agent behavior? To me, it doesn't seem like the LSTM adds much information about the past states.

Thanks for the help, and sorry about the many questions! I'm fairly new to reinforcement learning.

**Comment by richardliaw** [comment]
cc @ericl

**Comment by ericl** [comment]
What you're missing is the hidden state input and output of the LSTM. So a more complete diagram looks like this:
```
observation (o_t), hidden state (h_t)

                  ---------\
o_0, h_0         /   o_1, h_1                  (inputs)
 |    |         /     |     |
 |    |        /      |     |
model |       /      model  |
  \   /      /         \   /
   cell     /         cell -----.......
     |     /             |
    a_0, h_1            a_1, h_2               (outputs)

```

So the observation is first processed by the model, then the outputs of that and the LSTM hidden state are passed to the LSTM cell. The LSTM cell emits the new model outputs, and also the next hidden state.

The IMPALA paper has a nice figure:
![screenshot from 2018-08-01 15-50-00](https://user-images.githubusercontent.com/14922/43553173-92c9c958-95a2-11e8-89bc-3d73fa740f71.png)

**Comment by ericl** [comment]
Btw instead of using a LSTM, you can also use "preprocessor_pref": "deepmind" to enable framestacking. That works pretty well for atari.

**Comment by pschafhalter** [comment]
This clears a lot of things up, thanks for the explanation!

Is the `lstm_cell_size` the size of the hidden state?

**Comment by ericl** [comment]
That's right (see https://github.com/ray-project/ray/search?q=lstm_cell_size&unscoped_q=lstm_cell_size and https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/BasicRNNCell).

**Comment by pschafhalter** [comment]
Thanks for the help!

**Comment by LucCADORET** [comment]
Hi, sorry for digging this old issue. I am pretty new to machine learning in general, and this use_lstm option also bugs me. My main concern is why, with usual LSTM inputs, we input a 3D array (batch x time x features), but in all the exemples with use_lstm provided (such as https://github.com/ray-project/ray/blob/5d7afe8092f5521a8faa8cdfc916dcd9a5848023/rllib/examples/cartpole_lstm.py ), only a 2D array is provided.

**Comment by ericl** [comment]
RLlib records the episode lengths in a separate seq_lens tensor, and
automatically adds the time dimension based on the sequence lengths prior
to passing the data to the LSTM.

**Comment by adriendoerig** [comment]
Hi, sorry for reopening this. Your schematic from the 2nd of Aug 2018 is very helpful, but am I correct in assuming that the loss is backpropagated through time all the way through all the timesteps in an episode? Or is is only one timestep at a time? My task requires information integration over many timesteps, which requires the former. Thank you!

**Comment by ouissamam** [comment]
Hi, I also a question concerning the LSTM hidden states. I have the following setting: 'rollout_fragment_length': 32 and 'batch_mode': 'truncate_episodes' which results in a sample_batch with more than one episode in it. Will the hidden state be passed on through the whole sample_batch, or does rllib recognize that we have multiple episodes in one sample_batch and does not propagate the state across 2 different episodes? I ask this, because if it would be propagated across 2 episodes, we would have a temporal jump between 2 states.
